### ISSUE TITLE
Bug: ColumnChart columns halved if hAxis is date

### ISSUE THREAD

**Issue by orwant** [google/google-visualization-issues#1212]
```
This has been already reported, but it stil exists and is very annoying. Whenever I
draw a columnchart where the horizontal axis is date, the first and last columns are
halved, because their centers are positioned to the date. Since the hAxis ends with
the highest date it trims the first and last columns.

It would be great to automatically extend the hAxis in both directions by 1/2 barWidth.

PS: is there any way to make th first date's haxis value visible? As you can see only
the second column (and every other from there) is written, but not the first one.

thx a lot!

What steps will reproduce the problem? Please provide a link to a

http://jsfiddle.net/rokakoma/wEkjR/


What component is this issue related to (PieChart, LineChart, DataTable,
Query, etc)?

ColumnChart

Are you using the test environment (version 1.1)?
(If you are not sure, answer NO)

NO

What operating system and browser are you using?

All

*********************************************************
For developers viewing this issue: please click the 'star' icon to be
notified of future changes, and to let us know how many of you are
interested in seeing it resolved.
*********************************************************
```

Original issue reported on code.google.com by `rokakomma` on 2013-05-12 21:53:04

**Comment by orwant** [comment]
```
The problem is not just with date type but any continuous value.  For now, you can workaround
the problem by adding a viewWindow option with min and max that add half the barwidth.
 E.g. for your example:

           viewWindow: { 
              min: new Date(2012, 11, 15), 
              max: new Date(2013, 3, 15)
           }

Or you can avoid this problem by converting your values to strings so the axis will
use discrete values, which do get formatted correctly.
```

Original issue reported on code.google.com by `dlaliberte@google.com` on 2013-05-13 12:30:42

**Comment by orwant** [comment]
```
(No text was entered with this change)
```

Original issue reported on code.google.com by `dlaliberte@google.com` on 2013-05-13 12:31:18

**Comment by orwant** [comment]
```
Thanks, it's ALMOST perfect, but not :(

I need continous haxis, because I'm using a chartrangefilter as well. But if I set
hAxis.viewWindow.max to a certain date then chartrangefilter is not working anymore
properly, since whenever I track back the high end the chart itself still goes till
viewWindow.max even though the columns will disappear as they should after changing
the range.

If I set the 'range' for chartrangefilter for the extended period (adding 15days for
example) it doesn't have any effect.

If I set both viewWindow.max and state.range.end to the extended date it results the
same as in the first case, where the right end will always be the same date, just the
columns will disappear.

If I want chartrangefilter to work with columncharts, then I have to live with the
halved columns it seems now. Any other ideas?

Basically what I need is a hAxis.viewWindow.max to be in effect when the rangefilter
is at the high end, any other cases it should work as usually. The best would be if
chartrangefilter's state.range.end would always ovveride viewWindow.max, so by setting
only that value would solve all my problems as I would be able to set rangefilter overdated.

Thanks in advance!
```

Original issue reported on code.google.com by `rokakomma` on 2013-05-13 14:10:55

**Comment by orwant** [comment]
```
Use of a ColumnChart with the ChartRangeFilter is an excellent reason to fix this bug
with continuous values on the ColumnChart hAxis.  I can't think of any reasonable workaround.


Well, maybe after the ChartRangeFilter range is changed by the user, you could catch
the change, and programmatically change it again to a wider range.
```

Original issue reported on code.google.com by `dlaliberte@google.com` on 2013-05-13 14:26:11

**Comment by orwant** [comment]
```
Currently I'm working on to make it work with:

view.setColumns([0, {'calc': function(dataTable, rowIndex) { return dataTable.getFormattedValue(rowIndex,
0);},'type': 'string'}]);

This might solve my problem, but since I'm dynamically adding/hiding columns all the
time it makes a bit tough to regenerate the proper view each time, since I cannot set
1 single column only, I have to list all the columns, which differ in eery case. (I'm
working with ~600 different charts on my site :))

Yes, I was thinking about catching the 'statechange' which actually I'm already using
for storing the state, to redraw it properly after hiding/showing columns. But for
this case it seems an ugly hack.

Hopefully I'll solve it in the following half an hour, I'll let you know.
```

Original issue reported on code.google.com by `rokakomma` on 2013-05-13 15:01:07

**Comment by orwant** [comment]
```
Okay, I made it, it was a real pain:

So, at initialization:

var view   = new google.visualization.DataView(JSONdata);
var array  = view.getViewColumns();
array[0]={ 'calc': function(dataTable, rowIndex) { return dataTable.getFormattedValue(rowIndex,
0);},'type': 'string'};

wrapper.setView({'columns': array});
zeroview=array[0];   // needed later

### later when I hide/show columns

var view   = new google.visualization.DataView(JSONdata);
[...] // calculating the to show/hide columns, doing it with a series of view.hidecolumns();
var visiblecolumns=view.getViewColumns();
visiblecolumns[0]=zeroview; // needed to replace once again
wrapper.setView({'columns': visiblecolumns});  // not working with view.toJSON() even
though it's set with setColumns


------
The pain was, it wasn't working nor with 'setColumns' neither with DataView from JSON.
So even if I stored the original dataview, by var=zeroview=wrapper.getView(), or I
stored the entire array in zeroview (not just array[0]) it wasn't working.

For some mysterious reason it simply hid the first column (though it was in the array)
and made the second one to the hAxis.

Anyway, it's working know, so the chartrangefilter treats haxis values as a string,
so the first and last columns are entirely inside the viewWindow, and still rangefilter
is working fine.

Good news, haxis values got closer to eachother, since they became a string, which
I was unable to do before, so at least I got some additional improvement.

Thank you or your comments though, without knowing viewWindow exists I wouldn't have
started this little 'journey' at all :)

```

Original issue reported on code.google.com by `rokakomma` on 2013-05-13 16:52:40

**Comment by orwant** [comment]
```
Shoot, I have to correct myself, it's NOT working the way I'd like.

Since hAxis is treated as a string, chartranegefilter works almost fine, except when
I hide columns it doesn't shrink itself to show only that timewindow in which there
are actual data.

So if I have 2 dataseries, 1 from 2010-2012, the otherone from 2011-2013, if I throw
out the latter one, the hAxis still ends in 2013, but when it's treated as a date (continous)
it shrinks to 2012.

###
All in all, this won't work properly in any ways, unless chartrangefilter can be overdated.
:(
```

Original issue reported on code.google.com by `rokakomma` on 2013-05-13 17:00:03

**Comment by orwant** [comment]
```
I see two possible solutions:

a) overdate rangefilter, meaning if the highest datapoint is at Jan-1-2013 for example
I should be able to set the range till Jan-15-2013. (I can do this know, but it has
no effect)

b) if I set hAxis.viewWindow.min/max then chartrangefilter should automatically treat
these values as range max/min unless a tighter range is not defined.

b) seems pretty logical to me. 
```

Original issue reported on code.google.com by `rokakomma` on 2013-05-13 17:22:11

**Comment by orwant** [comment]
```
Overriding the ChartRangeFilter's ui.chartOptions.hAxis.min and .max should let you
do (a), but perhaps there are problems.

But I tend to agree that your (b) option also sounds reasonable.  The ChartRangeFilter
should be sensitive to the range of the chart it is filtering, not just the datatable
it sees as input.  Maybe there should be an option for that rather than making it the
default.
```

Original issue reported on code.google.com by `dlaliberte@google.com` on 2013-05-13 17:56:15

**Comment by orwant** [comment]
```
Thank you for your quick and detailed answers, I hope it will be solved one way or the
other. cheers, greg
```

Original issue reported on code.google.com by `rokakomma` on 2013-05-13 19:12:36

**Comment by orwant** [comment]
```
Meanwhile I figured out neither a) or b) options are really viable.

As I mentioned earlier, the problem arises when I hide columns. Generally, with a consinous
hAxis and no viewWindow specified, after hiding a column, the new viewWindow automatically
resizes itself to the new data boundaries.

But, if I define either a bigger hAxis.viewWindow or I convert hAxis values to string
(in the view only, as above) then the viewWindow doesn't resize itself anymore.

Resulting I will have empty area on the chart, where vAxis values are missing, but
hAxis values are still shown.

But if I want columnchart to include 1st/last columns entirely I can't hide columns
anymore, dynamically, as I can do it now with halved columns.

I'm not sure what the solution could be in this case. I will try to put up an example
to jsfiddle late afternoon, I just don't have time for this now.



```

Original issue reported on code.google.com by `rokakomma` on 2013-05-15 07:36:48

**Comment by orwant** [comment]
```
http://jsfiddle.net/rokakoma/ugDrw/

Here's the example. Sorry for being late, but I had my thesis deadline today :)

Anyway, if you try the above example, when you hide columns, the haxis automatically
resizes as it should (but columns are halved :( )

If you uncomment either viewindow, or viewmode as string, or both, then columns will
be included entirely but when you hide them haxis won't resize anymore.

What I'd like to achive, is to use haxis as a continous axis, with including entire
columns, so I could hide them and still get all of them included entirely.

I guess I can't use viewWindow for this purpose for sure. But for some reason then
haxis is viewed as a string it doesn't hide empty rows anymore, what it does with continous
haxis. And of course I still need chartrangefilter, but actually it works fine with
haxis viewed as string, but since the wrapper's viewwindow doesn't collapse neither
does rangefilter's


I hope I made my example clear enough, cheers!
```

Original issue reported on code.google.com by `rokakomma` on 2013-05-17 15:16:15

**Comment by orwant** [comment]
```
I may have been confusing, so in 1 word, what I'd like to have is the same behaviour,
as the above example does by default (resizing viewWindow) with the addition columns
are entirely included.

I think the simplest solution would be to let viewWindow be defined as a relative value
to hAxis range.

Something like:
hAxis: { 
  viewWindow: { 
    min: hAxis.minValue - new Date(0,0,15),
    max: hAxis.maxValue + new Date(0,0,15),
  }
}

Do you think it's possible? I assume options are evaluated each time draw() function
is called, meaning if you evaluate hAxis.minValue BEFORE viewWindow, then it would
work. It would also solve chartrangefilter's range as well, also would result the same
behaviour as I'd like to see.

Of coure I could genereate these values at every columnhide by checking all the rows,
but this seems a nice way to do.
```

Original issue reported on code.google.com by `rokakomma` on 2013-05-18 15:05:20

**Comment by orwant** [comment]
```
I have been working on a fix for this problem, but your last message suggests I should
(maybe) change this to account for whatever explicit minValue and maxValue have been
specified in the options.  My fix will first look for the smallest distance between
adjacent values, and that smallest gap size is then divided in half, which is subtracted
from the min and added to the max data value.

If a viewWindow min or max are specified, they will be used as is, so you would have
to account for the half gap size yourself.  
```

Original issue reported on code.google.com by `dlaliberte@google.com` on 2013-05-20 00:23:08

**Comment by orwant** [comment]
```
Sounds good for me, especially if it will work properly with the above example as well,
after dropping certain datacolumns.
Surely, if viewWindow is defined it shouldn't be overridden I agree, but hAxis.maxValue
is not always explicitly defined, yet it still exists, since there must be a minValue
anyway.

What I meant, was to use the hAxis.minvalue as the minvalue of the DataTable (or DataView
more likely). So when dataview changes so does minvalue (or at least it gets redefined)
even though it wasn't explicitly defined by the user. So when we hide datacolumns,
hAxis range might change. (but actually this is working fine now already)

Anyway, thank you for taking so much attention to this matter, I really appreciate
it.
```

Original issue reported on code.google.com by `rokakomma` on 2013-05-20 03:55:24
